src/java/org/apache/cassandra/locator/TokenMetadata.java

```diff
@@ -270,6 +270,41 @@ public void updateNormalTokens(Multimap<InetAddressAndPort, Token> endpointToken
         }
     }
 
+    /**
+     * Used by CNDB to update endpoint address for given normal tokens
+     */
+    public void updateAddressForNormalTokens(Collection<Token> tokens, InetAddressAndPort existing, InetAddressAndPort current)
+    {
+        assert tokens != null && !tokens.isEmpty();
+        assert existing != null && current != null;
+
+        lock.writeLock().lock();
+        try
+        {
+            Topology.Builder topologyBuilder = topology.unbuild();
+            bootstrapTokens.removeValue(current);
+            tokenToEndpointMap.removeValue(existing);
+            tokenToEndpointMap.removeValue(current);
+            topologyBuilder.addEndpoint(current);
+            leavingEndpoints.remove(current);
+            replacementToOriginal.remove(current);
+            removeFromMoving(current); // also removing this endpoint from moving
```

Comment:
On a closer look, shouldn't `remove` always be invoked with `existing`? Actually to reinforce the fact that this should only be invoked to replace fully joined nodes, should we even check `existing` is not part of bootstrapping/leaving nodes, failing fast otherwise?

Reply:
great point. Pushed 522f5cb8777 to add more validations to make sure `existing` has only normal tokens and `current` endpoint doesn't exist in TMD